### PR TITLE
Run clang++ in verbose mode during linking if swiftc is called with "-v"

### PR DESCRIPTION
**Pull request by dabelknap**
This PR is specific to `GenericUnix` and `Windows` platforms. If the `-v` option is supplied to the Swift compiler, also run `clang++` with the `-v` option during linking. Darwin calls the linker directly, so this change is only needed for Generic Unix and Windows.

Resolves [SR-656](https://bugs.swift.org/browse/SR-656).

Previously, this fix was attempted in [PR 4439](https://github.com/apple/swift/pull/4439), which was closed before any changes were made.

### CODE REVIEW

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by jrose-apple** [comment]
Hm. It still seems weird to me since we don't run anything *else* in verbose mode, but I guess Clang-as-linker is pretty well-understood. Okay.

**Comment by dabelknap** [comment]
@swift-ci please smoke test

**Comment by dabelknap** [comment]
@jrose-apple Do the smoke tests need to be re-triggered, or are they just taking a while?

**Comment by jrose-apple** [comment]
Oh, whoops, right.

@swift-ci Please smoke test

**Comment by dabelknap** [comment]
@jrose-apple Thanks for running the tests! I am a little confused by the Linux test failure. This wasn't an issue before I merged from `master`.

**Comment by jrose-apple** [comment]
:-( No, this one's a flaky test that I'm supposed to look into but haven't yet. Let's just run it again.

@swift-ci Please smoke test Linux

File: lib/Driver/UnixToolChains.cpp [hunk 1]
```diff
@@ -325,6 +325,11 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
   Arguments.push_back(
       context.Args.MakeArgString(context.Output.getPrimaryOutputFilename()));
 
+  // Run clang++ in verbose mode if "-v" is set
```
Comment (jrose-apple) [inline lib/Driver/UnixToolChains.cpp]: Oops, sorry. This violates the comment above! Please put the `-v` before the `-o` (in both implementations).

Comment (dabelknap) [reply lib/Driver/UnixToolChains.cpp]: Done.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Ugh, sorry. Is there a test you can add here? Even just one that combines `-v` and `-###` to show the `clang++` invocation.

**Comment by dabelknap** [comment]
I am working on writing the tests, but it is a little tricky since these changes are platform-dependent. Only Unix and Windows would be affected, since Darwin calls "ld" directly.

**Comment by jrose-apple** [comment]
You can specify an explicit target, which would force it to generate a command line for that platform. And then if you use `-###` / `-driver-print-jobs` (currently equivalent), it won't actually run the jobs.

File: test/Driver/verbose.swift [hunk 2]
```diff
@@ -0,0 +1,4 @@
+// RUN: %swiftc_driver -driver-print-jobs -target x86_64-unknown-linux-gnu -v %s 2>&1 | %FileCheck %s -check-prefix=VERBOSE_CLANG
+// RUN: %swiftc_driver -driver-print-jobs -target x86_64-unknown-windows-misc -v %s 2>&1 | %FileCheck %s -check-prefix=VERBOSE_CLANG
```
Comment (jrose-apple) [inline test/Driver/verbose.swift]: "misc"? Is this correct? (It probably doesn't matter for that part of the triple, but…)

**Review by jrose-apple — COMMENTED** [review]
Thanks, this looks good!

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

File: test/Driver/verbose.swift (hunk 2 above)
Comment (dabelknap) [reply test/Driver/verbose.swift]: That was probably a typo that I need to fix. That said, should it be "msvc" or "cygnus"?

Comment (jrose-apple) [reply test/Driver/verbose.swift]: "msvc", I think. I would expect the Cygnus build to use the GenericUnix toolchain. (Actually, there's a Cygnus toolchain that extends GenericUnix, but in this case it'll be the same.)

Comment (dabelknap) [reply test/Driver/verbose.swift]: Done.

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test and merge

**Comment by dabelknap** [comment]
Could we re-run the "Test and Merge" test? From looking at the log, it appears to have been an error relating to the GitHub API.

**Comment by jrose-apple** [comment]
When automation fails, do it by hand. Thanks for your patience on this one, Austin!

**Pull request merged**